runtime/memory.c

```diff
@@ -15,648 +15,347 @@
+      *p = newval;
+      write_barrier(obj, field, oldval, newval);
+      return 1;
+    } else {
+      return 0;
+    }
+  } else {
+    /* need a real CAS */
+    atomic_value* p = &Op_atomic_val(obj)[field];
+    if (atomic_compare_exchange_strong(p, &oldval, newval)) {
+      write_barrier(obj, field, oldval, newval);
+      return 1;
+    } else {
+      return 0;
     }
   }
-  CAMLassert (Wosize_hp (mem) >= request);
-  if (caml_add_to_heap ((char *) mem) != 0){
-    caml_free_for_heap ((char *) mem);
-    return NULL;
-  }
-  return Op_hp (mem);
 }
 
-/* Remove the heap chunk [chunk] from the heap and give the memory back
-   to [free].
-*/
-void caml_shrink_heap (char *chunk)
-{
-  char **cp;
-
-  /* Never deallocate the first chunk, because caml_heap_start is both the
-     first block and the base address for page numbers, and we don't
-     want to shift the page table, it's too messy (see above).
-     It will never happen anyway, because of the way compaction works.
-     (see compact.c)
-     XXX FIXME this has become false with the fix to PR#5389 (see compact.c)
-  */
-  if (chunk == caml_heap_start) return;
-
-  Caml_state->stat_heap_wsz -= Wsize_bsize (Chunk_size (chunk));
-  caml_gc_message (0x04, "Shrinking heap to %"
-                   ARCH_INTNAT_PRINTF_FORMAT "dk words\n",
-                   Caml_state->stat_heap_wsz / 1024);
 
-#ifdef DEBUG
-  {
-    mlsize_t i;
-    for (i = 0; i < Wsize_bsize (Chunk_size (chunk)); i++){
-      ((value *) chunk) [i] = Debug_free_shrink;
-    }
+CAMLprim value caml_atomic_load (value ref)
+{
+  if (caml_domain_alone()) {
+    return Field(ref, 0);
+  } else {
+    value v;
+    /* See Note [MM] above */
+    atomic_thread_fence(memory_order_acquire);
+    v = atomic_load(Op_atomic_val(ref));
+    return v;
   }
-#endif
-
-  -- Caml_state->stat_heap_chunks;
-
-  /* Remove [chunk] from the list of chunks. */
-  cp = &caml_heap_start;
-  while (*cp != chunk) cp = &(Chunk_next (*cp));
-  *cp = Chunk_next (chunk);
-
-  /* Remove the pages of [chunk] from the page table. */
-  caml_page_table_remove(In_heap, chunk, chunk + Chunk_size (chunk));
-
-  /* Free the [malloc] block that contains [chunk]. */
-  caml_free_for_heap (chunk);
 }
 
-CAMLexport color_t caml_allocation_color (void *hp)
+/* stores are implemented as exchanges */
+CAMLprim value caml_atomic_exchange (value ref, value v)
 {
-  if (caml_gc_phase == Phase_mark || caml_gc_phase == Phase_clean ||
-      (caml_gc_phase == Phase_sweep && (char *)hp >= (char *)caml_gc_sweep_hp)){
-    return Caml_black;
-  }else{
-    CAMLassert (caml_gc_phase == Phase_idle
-            || (caml_gc_phase == Phase_sweep
-                && (char *)hp < (char *)caml_gc_sweep_hp));
-    return Caml_white;
+  value ret;
+  if (caml_domain_alone()) {
+    ret = Field(ref, 0);
+    Field(ref, 0) = v;
+  } else {
+    /* See Note [MM] above */
+    atomic_thread_fence(memory_order_acquire);
+    ret = atomic_exchange(Op_atomic_val(ref), v);
```

Comment:
Here's a question about the implementation of the memory model again. For context, currently, `Atomic.store` and `Atomic.exchange` result directly in a call to `caml_atomic_exchange`.
Here if I follow naively the paper I expect instead something like:
```c
    ret = atomic_exchange(Op_atomic_val(ref), v);
    atomic_thread_fence(memory_order_release);
```
On armv8 (clang) the original code results in:
`dmb (ish)ld; L: ldaxr w2, [x0]; stlxr w3, w1, [x0]; cbnz w3, L` (notice the missing `dmb st`)
whereas the above one results in:
`L: ldaxr w2, [x0]; stlxr w3, w1, [x0]; cbnz w3, .L17; dmb (ish)`
which is closer to the paper (it does not seem to be possible to emit the `dmb st` barrier with standard C atomics; the closest is to use a full barrier).

Of course at the moment armv8 is not supported (but I assume that the armv8 prototype is currently implemented with these functions). However this runs deeper:
* The implementation currently seems correct on x86 when these functions are called from OCaml. However one could imagine that when they are called from C code, a compiler might try to inline them and be too smart. How do you ensure that these functions implement the OCaml memory model on x86 when called from C?
* Maybe the memory model has evolved between the first prototype and the paper and the documentation at the top of the file was not updated. In that case the comment should it be updated (more that I initially thought).

Naively I would think that for the OCaml-C FFI you are eventually going to implement a conservative approximation of the OCaml memory model in C, whereas for native OCaml code you will eventually emit the exact assembly code you need (plus a C call to the write barrier). Sorry if there is a simple answer that I am missing!